### PR TITLE
Feature/tr 1867/launch test with lti 1p3

### PR DESCRIPTION
**Pull request by emgolubev**
Related to: https://oat-sa.atlassian.net/browse/TR-1867

Test:
- you can use https://github.com/oat-sa/nextgen-stack/tree/feature/TR-1577/oidc-initiation for easier testing

Example

http://lti.docker.localhost/platform/message/launch/lti-resource-link?registration=devkitToTAO&user_list=c3po&launch_url=http://backoffice.docker.localhost/ltiDeliveryProvider/DeliveryTool/launch1p3/?delivery%3Dhttp%253A%252F%252Fbackoffice.docker.localhost%252Fontologies%252Ftao.rdf%2523i613888211402388794933c0190e372&claims=%7B%0D%0A%20%20%20%20%22https://purl.imsglobal.org/spec/lti/claim/roles%22:%20%5B%0D%0A%20%20%20%20%20%20%20%20%22http://purl.imsglobal.org/vocab/lis/v2/membership%23Learner%22%0D%0A%20%20%20%20%5D%0D%0A%7D

### CODE REVIEW

File: models/classes/LtiService.php [hunk 1]
```diff
@@ -61,6 +64,27 @@ public function createLtiSession(common_http_Request $request)
...
+
+            $this->getServiceLocator()->propagate($session);
+
+            return $session;
+        } catch (LtiInvalidVariableException $e) {
+            $this->getServiceLocator()->get(LoggerService::SERVICE_ID)
+                ->log(LogLevel::INFO, $e->getMessage());
+
+            throw new LtiException(
+                __($e->getMessage()),
```
Comment (alexeyvy) [inline models/classes/LtiService.php:82]: I don't think we want exceptions translated, so __() can be deleted

File: models/classes/user/Lti1p3User.php [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+ *
+ * @access public
+ * @author Joel Bout, <joel@taotesting.com>
+ * @package taoLti
+ */
+class Lti1p3User extends \common_user_User implements ServiceLocatorAwareInterface, \JsonSerializable, LtiUserInterface
+{
+    use ServiceLocatorAwareTrait;
+
+    const USER_IDENTIFIER = 'identifier';
```
Comment (alexeyvy) [inline models/classes/user/Lti1p3User.php:46]: let's add an access modifier here

File: models/classes/user/Lti1p3User.php [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+     * @throws \oat\taoLti\models\classes\LtiVariableMissingException
+     */
+    public function __construct($launchData)
+    {
+        $this->ltiLaunchData = $launchData;
+        $this->userUri = $launchData->getVariable(LtiLaunchData::USER_ID) ?? 'anonymous';
+
+        $taoRoles = $this->determineTaoRoles($launchData);
+        if (empty($taoRoles)) {
+            $message = "Invalid LTI role parameter value: " . $this->ltiLaunchData->getVariable(LtiLaunchData::ROLES);
```
Comment (alexeyvy) [inline models/classes/user/Lti1p3User.php:96]:
```suggestion
            $message = 'Invalid LTI role parameter value: ' . $this->ltiLaunchData->getVariable(LtiLaunchData::ROLES);
```

File: composer.json [hunk 4]
```diff
@@ -63,8 +63,12 @@
     },
     "autoload" : {
         "psr-4" : {
-            "oat\\taoLti\\" : "",
-            "OAT\\Library\\Lti1p3Core\\Tests\\": "../vendor/oat-sa/lib-lti1p3-core/tests/"
+            "oat\\taoLti\\" : ""
         }
+    },
+    "autoload-dev" : {
```
Comment (alexeyvy) [inline composer.json:69]: this was added as part of the previous PR by Mathieu. Not sure why this diff is in this PR as well, something wrong with branches?

File: install/ontology/ltiroles_membership.rdf [hunk 5]
```diff
@@ -50,4 +50,12 @@
     <generis:includesRole rdf:resource="http://www.tao.lu/Ontologies/TAOLTI.rdf#LtiBaseRole"/>
     <generis:isSystem rdf:resource="http://www.tao.lu/Ontologies/generis.rdf#True"/>
   </rdf:Description>
+  <rdf:Description rdf:about="http://purl.imsglobal.org/vocab/lis/v2/membership#Learner">
```
Comment (alexeyvy) [inline install/ontology/ltiroles_membership.rdf:53]: We need a migration with syncModels() for this change to take effect on existing instances

File: composer.json (hunk 4 above)
Comment (emgolubev) [reply composer.json:69]: Do not see https://github.com/oat-sa/extension-tao-lti/pull/300/files

File: install/ontology/ltiroles_membership.rdf (hunk 5 above)
Comment (emgolubev) [reply install/ontology/ltiroles_membership.rdf:53]: We have one in ltiDeliveryProvider, should I duplicate it?

File: models/classes/TaoLti1p3Session.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Returns an resource representing the incoming link
+     *
+     * @return core_kernel_classes_Resource
+     * @throws LtiVariableMissingException
+     * @throws common_exception_Error
+     */
+    public function getLtiLinkResource()
+    {
+        if (is_null($this->ltiLink)) {
```
Comment (wazelin) [inline models/classes/TaoLti1p3Session.php:51]:
```suggestion
        if (null === $this->ltiLink) {
```

File: models/classes/user/Lti1p3User.php [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+        $this->firstname = $firstname;
+        $this->lastname = $lastname;
+        $this->email = $email;
+        $this->label = $label;
+    }
+
+    public function setRoles($roles)
+    {
+        $newRoles = array_map(function ($value) {
```
Comment (wazelin) [inline models/classes/user/Lti1p3User.php:121]:
```suggestion
        $newRoles = array_map(static function ($value) {
```

File: models/classes/user/Lti1p3User.php [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+use oat\oatbox\user\UserLanguageService;
+
+/**
+ * Authentication adapter interface to be implemented by authentication methodes
+ *
+ * @access public
+ * @author Joel Bout, <joel@taotesting.com>
+ * @package taoLti
+ */
+class Lti1p3User extends \common_user_User implements ServiceLocatorAwareInterface, \JsonSerializable, LtiUserInterface
```
Comment (wazelin) [inline models/classes/user/Lti1p3User.php:42]: Any chance to reuse `LtiUser`'s methods here by either
1. Extracting the common ones into dedicated services (e.g. a serializer)
2. Including the LtiUser object as a dependency of this new class, delegating some method calls to the internal object, having sort of a composition of both.

**Review by wazelin — CHANGES_REQUESTED** [review]
Very good job, @emgolubev!

I'm just a bit concerned by the factually duplicated LtiUser domain methods. If that could be easily fixed, it'd be awesome.

File: models/classes/user/Lti1p3User.php (hunk 8 above)
Comment (emgolubev) [reply models/classes/user/Lti1p3User.php]: done

File: models/classes/TaoLti1p3Session.php [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+//            $service = $this->getServiceLocator()->get(LinkService::SERVICE_ID);
+//            $consumer = new \core_kernel_classes_Resource('eugene'); //$this->getLaunchData()->getLtiConsumer();
+//            $linkId = $service->getLinkId($consumer->getUri(), $this->getLaunchData()->getResourceLinkID());
+
+            // What lti link is? What purpose?
```
[the remark this answers is not in the document]
Comment (emgolubev) [reply models/classes/TaoLti1p3Session.php]: changed, please review

File: models/classes/user/Lti1p3User.php [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * Calculate your primary tao roles from the launch data
+     *
+     * @param LtiLaunchData $ltiLaunchData
+     * @return array
+     * @throws \common_Exception
+     * @throws \common_exception_Error
+     */
+    protected function determineTaoRoles(LtiLaunchData $ltiLaunchData)
```
Comment (alexeyvy) [inline models/classes/user/Lti1p3User.php:61]: I see it mostly duplicates the parent method's code, could we re-use it?

Comment (emgolubev) [reply models/classes/user/Lti1p3User.php:61]: already done

File: models/classes/TaoLti1p3Session.php [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+use core_kernel_classes_Resource;
+use oat\taoLti\models\classes\ResourceLink\LinkService;
+
+/**
+ * The TAO layer ontop of the LtiSession
+ *
+ * @access public
+ * @author Joel Bout, <joel@taotesting.com>
+ * @package taoLti
+ */
```
Comment (augustas) [inline models/classes/TaoLti1p3Session.php:35]:
```suggestion
```
I don't see much use of this class phpdoc, if you disagree please update it :)

File: models/classes/user/Lti1p3User.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+use oat\taoLti\models\classes\LtiRoles;
+use oat\taoLti\models\classes\LtiUtils;
+use oat\taoLti\models\classes\LtiVariableMissingException;
+
+/**
+ * Authentication adapter interface to be implemented by authentication methodes
+ *
+ * @access public
+ * @package taoLti
+ */
```
Comment (augustas) [inline models/classes/user/Lti1p3User.php:35]:
```suggestion
```
Feel free to disagree and update

File: migrations/Version202109040100000000_taoLti.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+declare(strict_types=1);
+
+namespace oat\taoLti\migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\Exception\IrreversibleMigration;
+use oat\tao\scripts\SyncModels;
+use oat\tao\scripts\tools\migrations\AbstractMigration;
+
+final class Version202109040100000000_taoLti extends AbstractMigration
```
Comment (augustas) [inline migrations/Version202109040100000000_taoLti.php:12]: Seems to be a custom version number, if you'll need generate another migration you can run f.e.: `docker exec -it TAO_PHPFPM_CONTAINER php index.php 'oat\tao\scripts\tools\Migrations' -c generate -e taoLti`

**Review by augustas — APPROVED** [review]
Looking good, works fine locally. Mapping from `LtiMessagePayload` to `LtiLaunchData::$variables` could be tested. 

- [ ] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

**Review by alexeyvy — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful

File: models/classes/LtiLaunchData.php [hunk 14]
```diff
@@ -124,6 +127,38 @@ public static function fromRequest(common_http_Request $request)
...
+        $variables[self::LAUNCH_PRESENTATION_RETURN_URL] = $ltiMessagePayload->getLaunchPresentation() ? $ltiMessagePayload->getLaunchPresentation()->getReturnUrl() : null;
+        $variables[self::LTI_VERSION] = $ltiMessagePayload->getVersion();
+        $variables[self::LTI_MESSAGE_TYPE] = $ltiMessagePayload->getMessageType();
+        $variables[self::LIS_RESULT_SOURCEDID] = $ltiMessagePayload->getBasicOutcome() ? $ltiMessagePayload->getBasicOutcome()->getLisResultSourcedId() : null;
+        $variables[self::LIS_OUTCOME_SERVICE_URL] = $ltiMessagePayload->getBasicOutcome() ? $ltiMessagePayload->getBasicOutcome()->getLisOutcomeServiceUrl() : null;
+
+        if ($platform) {
+            $variables[self::TOOL_CONSUMER_INSTANCE_ID] = $platform->getLabel();
+            $variables[self::TOOL_CONSUMER_INSTANCE_NAME] = $platform->getLabel();
+            $variables[self::TOOL_CONSUMER_INSTANCE_DESCRIPTION] = $platform->getLabel();
```
[the remark this answers is not in the document]
Comment (emgolubev) [reply models/classes/LtiLaunchData.php]: fixed

File: test/unit/models/classes/Security/AuthorizationServerFactoryTest.php [hunk 15]
```diff
@@ -42,8 +42,8 @@ class AuthorizationServerFactoryTest extends TestCase
 {
     use NoPrivacyTrait;
 
-    private const PRIVATE_KEY_PATH = __DIR__ . '/../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/private.key';
-    private const PUBLIC_KEY_PATH = __DIR__ . '/../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/public.key';
+    private const PRIVATE_KEY_PATH = __DIR__ . '/../../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/private.key';
+    private const PUBLIC_KEY_PATH = __DIR__ . '/../../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/public.key';
```
Comment (wazelin) [inline test/unit/models/classes/Security/AuthorizationServerFactoryTest.php:46]: This causes the CI pipeline error.
```
1) unit\models\classes\Security\AuthorizationServerFactoryTest::testCreate

file_get_contents(/var/www/test/unit/models/classes/Security/../../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/private.key): failed to open stream: No such file or directory
```
Let's put it back :)
```suggestion
    private const PRIVATE_KEY_PATH = __DIR__ . '/../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/private.key';
    private const PUBLIC_KEY_PATH = __DIR__ . '/../../../../../vendor/oat-sa/lib-lti1p3-core/tests/Resource/Key/RSA/public.key';
```

**Review by wazelin — CHANGES_REQUESTED** [review]

**Review by wazelin — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

**Pull request merged**